---
title: "BoneCellType: R package for bone-relevant cell type assignment"
author: Kristyna Kupkova
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    theme: flatly # many options for theme, this one is my favorite.
    highlight: pygments
    df_print: paged
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{BoneCellType: R package for bone-relevant cell type assignment}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r klippy, echo=FALSE, include=TRUE}
# add copy clipboard button
klippy::klippy(position = "right")
```

```{r, echo=FALSE}
# These settings make the vignette prettier
knitr::opts_chunk$set(results="hold", collapse=FALSE, message=FALSE, 
                      warning = FALSE)
```

# Introduction to BoneCellType

The BoneCellType package aims to help assign cell types to single-cell
RNA-seq clusters derived from bone samples. BoneCellType uses a
custom-assembled database of bone-relevant markers (assembly approach
described [here](https://github.com/kkupkova/Mouse-bone-markers)). There
are two approaches introduced here to reach this goal: 1) marker
overrepresentation analysis using Fisher's exact test, and 2) gene
(marker) set enrichment analysis.

**Note!** The results produced with BoneCellType do not assign
definitive cell-type labels; only an initial prediction based on
statistical testing and further manual curation is needed.

# Setup

Let's load the BoneCellType package along with a few other packages
whose functions are used in this tutorial.

The input into the functions presented here is a data.frame with
variable markers (i.e. features) for individual clusters that can be
obtained by using [Seurat](https://satijalab.org/seurat/) `FindMarkers`
function. Here we provide a previously generated example dataset.

```{r setup}
library(BoneCellType)
library(dplyr)
library(ggplot2)

# example dataset
varMarkerFile = system.file("extdata", "cluster_variable_features.csv", 
                            package = "BoneCellType")
varMarkers = read.csv(varMarkerFile, sep = ",")

# show top rows of the example dataset
head(varMarkers)
```

# 1. Marker overrepresentation analysis

To calculate marker overrepresentation use `markerEnrich` function. It
automatically provides a database of bone-relevant markers. Results can
be further dirrectly plotted with an associated `plotMarkerEnrich`
function, output of which is an editable *ggplot* object.

**Note**: While the default setting of `markerEnrich` use all available
variable markers per cluster, based on practical experience, more
reliable results can be achieved by selecting only few top markers by
setting the parameter *topN* (as shown in the example below). Default
mode selects the top markers based on log2FC, this can be changed by
setting the parameter *sortBy="padj"*. Changing the *sortBy*, however,
usually does not change much.

```{r marker-overrep-top10, fig.width=6, fig.height=6}
# run overrepresentation analysis uasing top 10 variable markers (features) per cluster
ORres = markerEnrich(varMarkers, topN = 10)

# plot the top 5 most significant results
plotMarkerEnrich(ORres, topN = 5)
```

The labels on the x-axis contain following information \<cluster number
\_\_ cell type \_\_ source database\>. Dashed line indicates padj =
0.05.

Let's now have a look at the format of the `markerEnrich` results.

```{r topRes}
# show top 2 significant result for each cluster
topORres = ORres %>% 
  group_by(cluster) %>% 
  top_n(n = 2, wt = neg_log10_padj)

topORres
```

As shown, the output is a table with following variables- *cluster*:
cluster number / *cellType*: cell type / *database*: source database
from which markers originate / *p*: p-value / *oddsRatio*: odds ratio /
*n_olaps*: number of overlap between variable markers (features) and the
marker database / *overlaps*: list of individual overlapping markers /
*padj*: adjusted p-value /*neg_log10_padj*: -log10(padj).

**Recommendation:** Try few different settings of the `markerEnrich`
function try different *topN*, *sortBy* arguments. In general I wouldn't
set the *byDatabase* to FALSE, as the TISSUES database has relatively a
lot of markers (some of which I am not too sure about) and it can skew
the results. But overall I definitely recommend to explore the whole
result table, to see which markers drive associated results and if that
cell type assignment then makes sense.

# 2. Gene (marker) set enrichment analysis

To perfeorm GSEA analysis (with the use of [fgsea
package](https://bioconductor.org/packages/devel/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html))
use `markerGSEA` function. Similarly to `markerEnrich`a, this function
also automatically provides a database of bone-relevant markers. Results
can then be plotted with `plotMarkerGSEA` function, output of which is
an editable *ggplot* object.

Unlike overrepresentationa analysis, GSEA analysis (`markerGSEA` ) uses
full list of markers associated with a given cluster sorted by
log~2~(fold change) values. Note that the list of variable markers
(features) per cluster is usually not very long, therefore
overrepresentation analysis (introduced above) might be more suitable.

```{r marker-GSEA, fig.width=6, fig.height=6}
# run GSEA
GSEAres = markerGSEA(varMarkers)

# plot the top 5 most significant results
plotMarkerGSEA(GSEAres, topN = 5)
```

The labels on the x-axis contain following information \<cluster number
\_\_ source database \_\_ cell type\>. Dashed line indicates padj =
0.05.

Let's now have a look at the format of the `markerGSEA` results.

```{r topResGSEA}
# show top 2 significant result for each cluster
topGSEAres = GSEAres %>% 
  group_by(cluster) %>% 
  top_n(n = 2, wt = -padj)

topGSEAres
```

As shown, the output is a table with following variables- *pathaway*:
defines a set of markers used for GSEA analysis / *pval*: p-value /
*padj*: adjusted p-value / *log2err*: the expected error for the
standard deviation of the p-value logarithm / *ES*: enrichment score /
*NES*: normalized enrichment score /*size*: size of the pathway after
removing genes not present in 'names(stats)' / *leadingEdge*: vector
with indexes of leading edge genes that drive the enrichment /*cluster*:
cluster number.

**Note:** As you can see in the plot above, the results in cluster 2
have very low significance in terms of adjusted p-value. In this case it
might be useful to plot *NES* values instead. Let's have a look how to
do that.

```{r plot-NES, fig.width=6, fig.height=6}
# plot the top 5 most significant results
plotMarkerGSEA(GSEAres, topN = 5, plotNES = TRUE)
```

# Custom settings

The functions presented here provide multiple options how to run
analyses and plot the results. Additionally, users can provide their own
sets of cell type / tissue markers (these do not have to be
bone-related). Furthermore, the outputs of the plotting function are
editable *ggplot* objects. In this section we will cover both of these
options to customize provided functions.

## a. Provide custom set of markers

As part pf the package we provide a set of bone-related markers. The
inputs to overrepresentation analysis vs. GSEA analysis have slightly
different format (*data.frame* vs. *list*, respectively). So let's have
a look at what format are the inputs and how do we provide those to both
functions from this package.

### a.1 Markers for overrepresentation analysis

The marker database provided in this package for overrepresentation
analysis can be accessed in following way:

```{r overrepMarkers}
# access set of markers
markerDatabase = markerDatabase

# view variables in the markerDatabase data.frame
head(markerDatabase)
```

If you want to provide your own set of markers, the *data.frame* must
have all following columns: *tissue, cellType, gene, database*. Each row
must contain only one gene (i.e. marker). Let's now have a look how to
use the markerDatabase object in the overrepresentation analysis with
`markerEnrich`.

```{r overrepCutomMarkers, fig.width=6, fig.height=6}
# run overrepresentation analysis uasing top 10 variable markers (features) per cluster
ORres2 = markerEnrich(varMarkers = varMarkers, markerDB = markerDatabase, topN = 10)

# plot the top 5 most significant results
plotMarkerEnrich(ORres2, topN = 5)
```

Using the input in this way allows us to for example filter out markers
coming from sources we might not trust as much. Let's say we would like
to omit markers coming from *CellMarker2_mouse_sc* database (source).
Let's have a look how to do that.

```{r overrepFilter, fig.width=6, fig.height=6}
# filter out markers originating from CellMarker2_mouse source
filtDB = markerDatabase %>% 
  dplyr::filter(database != "CellMarker2_mouse_sc")
  
# pass the filtered set of markers to overrepresentation analysis
ORresFilt = markerEnrich(varMarkers = varMarkers, markerDB = filtDB, topN = 10)

# plot the top 5 most significant results
plotMarkerEnrich(ORresFilt, topN = 5)
```

As you can see, cell types from *CellMarker2_mouse_sc* are not reported
in the results, since the markers were filtered out.

### a.2 Markers for GSEA analysis

Marker input to the GSEA analysis is in form of a named *list*. The
names correspond to a given set and the content of the given set is a
string vector containing given markers. There are 4 marker sets included
in this package, let's have a look at them.

1.  **List with cell type specific markers merged from multiple source
    databases**

```{r GSEAcellMarkers}
# access set of cell type markers (merged from multiple source databases)
cellType_list = cellType_list

# view contents (maximum 6 markers per sublist)
headList(cellType_list)
```

2.  **List with tissue specific markers merged from multiple source
    databases**

```{r GSEAtissueMarkers}
# access set of tissue markers (merged from multiple source databases)
tissue_list = tissue_list

# view contents (maximum 6 markers per sublist)
headList(tissue_list)
```

3.  **List with cell type specific markers separated by source
    database**

```{r GSEAcellSourceMarkers}
# access set of cell type markers separated by source
cellTypeSource_list = cellTypeSource_list

# view contents (maximum 6 markers per sublist)
headList(cellTypeSource_list)
```

4.  **List with cell type specific markers merged from multiple source
    databases**

```{r GSEAtissueSourceMarkers}
# access set of tissue markers separated by source
tissueSource_list = tissueSource_list

# view contents (maximum 6 markers per sublist)
headList(tissueSource_list)
```

Now let's pass a selected set of markers to the markerGSEA function.
**Note:** once you provide your list of markers to the *markers*
variable, *systemLevel* and *byDatabase* options will not do anything
anymore. You have to create your list according to your needs.

```{r marker-GSEA-custom, fig.width=6, fig.height=6}
# run GSEA with your set of markers
GSEAcellTypeSource = markerGSEA(varMarkers = varMarkers, markers = cellTypeSource_list)

# plot NES of the top 5 most significant results
plotMarkerGSEA(GSEAcellTypeSource, topN = 5, plotNES = TRUE)
```

## b. Edit generated plots

The outputs of the plotting functions are editable *ggplot* objects that
you can adjust to your liking. Let's have a look how to do that.

First let's generate the default plot.

```{r marker-overrep, fig.width=6, fig.height=6}
# run overrepresentation analysis uasing top 10 variable markers (features) per cluster
ORres = markerEnrich(varMarkers, topN = 10)

# plot the top 5 most significant results
p = plotMarkerEnrich(ORres, topN = 5)
p
```

Let's now change the color scheme and add a plot title.

```{r edit plot, fig.width=6, fig.height=6}
library(viridisLite)
p = p + 
  scale_fill_viridis_c(option = "D", direction = -1) +
  ggtitle("Bone cell type overrepresentation analysis")
p
```
