<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BoneCellType">
<title>BoneCellType: R package for bone-relevant cell type assignment • BoneCellType</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="BoneCellType: R package for bone-relevant cell type assignment">
<meta property="og:description" content="BoneCellType">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BoneCellType</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/intro_to_BoneCellType.html">BoneCellType: R package for bone-relevant cell type assignment</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="intro_to_BoneCellType_files/clipboard-1.7.1/clipboard.min.js"></script><link href="intro_to_BoneCellType_files/primer-tooltips-1.4.0/build.css" rel="stylesheet">
<link href="intro_to_BoneCellType_files/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet">
<script src="intro_to_BoneCellType_files/klippy-0.0.0.9500/js/klippy.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>BoneCellType: R package for bone-relevant cell type assignment</h1>
                        <h4 data-toc-skip class="author">Kristyna
Kupkova</h4>
            
            <h4 data-toc-skip class="date">2023-10-03</h4>
      
      
      <div class="d-none name"><code>intro_to_BoneCellType.Rmd</code></div>
    </div>

    
    
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script><div class="section level2">
<h2 id="introduction-to-bonecelltype">Introduction to BoneCellType<a class="anchor" aria-label="anchor" href="#introduction-to-bonecelltype"></a>
</h2>
<p>The BoneCellType package aims to help assign cell types to
single-cell RNA-seq clusters derived from bone samples. BoneCellType
uses a custom-assembled database of bone-relevant markers (assembly
approach described <a href="https://github.com/kkupkova/Mouse-bone-markers" class="external-link">here</a>). There
are two approaches introduced here to reach this goal: 1) marker
overrepresentation analysis using Fisher’s exact test, and 2) gene
(marker) set enrichment analysis.</p>
<p><strong>Note!</strong> The results produced with BoneCellType do not
assign definitive cell-type labels; only an initial prediction based on
statistical testing and further manual curation is needed.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Let’s load the BoneCellType package along with a few other packages
whose functions are used in this tutorial.</p>
<p>The input into the functions presented here is a data.frame with
variable markers (i.e. features) for individual clusters that can be
obtained by using <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>
<code>FindMarkers</code> function. Here we provide a previously
generated example dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://farber-lab.github.io/BoneCellType/">BoneCellType</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># example dataset</span></span>
<span><span class="va">varMarkerFile</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"cluster_variable_features.csv"</span>, </span>
<span>                            package <span class="op">=</span> <span class="st">"BoneCellType"</span><span class="op">)</span></span>
<span><span class="va">varMarkers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">varMarkerFile</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show top rows of the example dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">varMarkers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;           p_val avg_log2FC pct.1 pct.2     p_val_adj cluster   gene</span></span>
<span><span class="co">#&gt; 1 1.551369e-202   7.978416 0.664 0.021 2.861810e-198       0  Hemgn</span></span>
<span><span class="co">#&gt; 2 5.127751e-104  16.393282 0.935 0.169 9.459163e-100       0  Mki67</span></span>
<span><span class="co">#&gt; 3  4.012367e-98   5.801448 0.701 0.078  7.401614e-94       0 Diaph3</span></span>
<span><span class="co">#&gt; 4  4.156995e-96  11.168982 0.776 0.107  7.668409e-92       0   Ank1</span></span>
<span><span class="co">#&gt; 5  1.869814e-87   8.421892 0.776 0.120  3.449246e-83       0   Sptb</span></span>
<span><span class="co">#&gt; 6  1.936041e-87   7.841105 0.785 0.123  3.571415e-83       0  Top2a</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="marker-overrepresentation-analysis">1. Marker overrepresentation analysis<a class="anchor" aria-label="anchor" href="#marker-overrepresentation-analysis"></a>
</h2>
<p>To calculate marker overrepresentation use <code>markerEnrich</code>
function. It automatically provides a database of bone-relevant markers.
Results can be further dirrectly plotted with an associated
<code>plotMarkerEnrich</code> function, output of which is an editable
<em>ggplot</em> object.</p>
<p><strong>Note</strong>: While the default setting of
<code>markerEnrich</code> use all available variable markers per
cluster, based on practical experience, more reliable results can be
achieved by selecting only few top markers by setting the parameter
<em>topN</em> (as shown in the example below). Default mode selects the
top markers based on log2FC, this can be changed by setting the
parameter <em>sortBy=“padj”</em>. Changing the <em>sortBy</em>, however,
usually does not change much.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run overrepresentation analysis uasing top 10 variable markers (features) per cluster</span></span>
<span><span class="va">ORres</span> <span class="op">=</span> <span class="fu"><a href="../reference/markerEnrich.html">markerEnrich</a></span><span class="op">(</span><span class="va">varMarkers</span>, topN <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the top 5 most significant results</span></span>
<span><span class="fu"><a href="../reference/plotMarkerEnrich.html">plotMarkerEnrich</a></span><span class="op">(</span><span class="va">ORres</span>, topN <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/marker-overrep-top10-1.png" width="576"></p>
<p>The labels on the x-axis contain following information &lt;cluster
number __ cell type __ source database&gt;. Dashed line indicates padj =
0.05.</p>
<p>Let’s now have a look at the format of the <code>markerEnrich</code>
results.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show top 2 significant result for each cluster</span></span>
<span><span class="va">topORres</span> <span class="op">=</span> <span class="va">ORres</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, wt <span class="op">=</span> <span class="va">neg_log10_padj</span><span class="op">)</span></span>
<span></span>
<span><span class="va">topORres</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   cluster [3]</span></span></span>
<span><span class="co">#&gt;   cluster cellType           database       p oddsRatio n_olaps overlaps    padj</span></span>
<span><span class="co">#&gt;     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>       0 Macrophage         CellMar… 1.05<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>     169.        2 Mki67,T… 2.20<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>       0 Erythroblast       MSigDB   7.99<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>      48.4       3 Top2a,C… 2.20<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>       1 Fibroblast         CellMar… 1.32<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>     515.        2 Col1a2,… 8.31<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>       1 Preosteoblast      CellMar… 9.43<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>    <span style="text-decoration: underline;">2</span>339.        1 Col1a1   9.90<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>       2 Central memory T … CellMar… 2.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>     574.        1 Lef1     1.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>       2 Central memory T … CellMar… 2.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>     574.        1 Lef1     1.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: neg_log10_padj &lt;dbl&gt;</span></span></span></code></pre>
<p>As shown, the output is a table with following variables-
<em>cluster</em>: cluster number / <em>cellType</em>: cell type /
<em>database</em>: source database from which markers originate /
<em>p</em>: p-value / <em>oddsRatio</em>: odds ratio / <em>n_olaps</em>:
number of overlap between variable markers (features) and the marker
database / <em>overlaps</em>: list of individual overlapping markers /
<em>padj</em>: adjusted p-value /<em>neg_log10_padj</em>:
-log10(padj).</p>
<p><strong>Recommendation:</strong> Try few different settings of the
<code>markerEnrich</code> function try different <em>topN</em>,
<em>sortBy</em> arguments. In general I wouldn’t set the
<em>byDatabase</em> to FALSE, as the TISSUES database has relatively a
lot of markers (some of which I am not too sure about) and it can skew
the results. But overall I definitely recommend to explore the whole
result table, to see which markers drive associated results and if that
cell type assignment then makes sense.</p>
</div>
<div class="section level2">
<h2 id="gene-marker-set-enrichment-analysis">2. Gene (marker) set enrichment analysis<a class="anchor" aria-label="anchor" href="#gene-marker-set-enrichment-analysis"></a>
</h2>
<p>To perfeorm GSEA analysis (with the use of <a href="https://bioconductor.org/packages/devel/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html" class="external-link">fgsea
package</a>) use <code>markerGSEA</code> function. Similarly to
<code>markerEnrich</code>a, this function also automatically provides a
database of bone-relevant markers. Results can then be plotted with
<code>plotMarkerGSEA</code> function, output of which is an editable
<em>ggplot</em> object.</p>
<p>Unlike overrepresentationa analysis, GSEA analysis
(<code>markerGSEA</code> ) uses full list of markers associated with a
given cluster sorted by log<sub>2</sub>(fold change) values. Note that
the list of variable markers (features) per cluster is usually not very
long, therefore overrepresentation analysis (introduced above) might be
more suitable.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run GSEA</span></span>
<span><span class="va">GSEAres</span> <span class="op">=</span> <span class="fu"><a href="../reference/markerGSEA.html">markerGSEA</a></span><span class="op">(</span><span class="va">varMarkers</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the top 5 most significant results</span></span>
<span><span class="fu"><a href="../reference/plotMarkerGSEA.html">plotMarkerGSEA</a></span><span class="op">(</span><span class="va">GSEAres</span>, topN <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/marker-GSEA-1.png" width="576"></p>
<p>The labels on the x-axis contain following information &lt;cluster
number __ source database __ cell type&gt;. Dashed line indicates padj =
0.05.</p>
<p>Let’s now have a look at the format of the <code>markerGSEA</code>
results.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show top 2 significant result for each cluster</span></span>
<span><span class="va">topGSEAres</span> <span class="op">=</span> <span class="va">GSEAres</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, wt <span class="op">=</span> <span class="op">-</span><span class="va">padj</span><span class="op">)</span></span>
<span></span>
<span><span class="va">topGSEAres</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 33 × 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   cluster [3]</span></span></span>
<span><span class="co">#&gt;    pathway             pval  padj log2err     ES   NES  size leadingEdge cluster</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> CellMarker2_mous… 0.094<span style="text-decoration: underline;">2</span> 0.363   0.170 -<span style="color: #BB0000;">0.905</span> -<span style="color: #BB0000;">1.42</span>     7 Hba-a1            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> CellMarker2_mous… 0.168  0.363   0.127 -<span style="color: #BB0000;">0.852</span> -<span style="color: #BB0000;">1.33</span>     6 S100a9,Ltf…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> CellMarker2_mous… 0.073<span style="text-decoration: underline;">8</span> 0.363   0.190 -<span style="color: #BB0000;">0.901</span> -<span style="color: #BB0000;">1.44</span>     8 Hba-a1,Hbb…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> CellMarker2_mous… 0.081<span style="text-decoration: underline;">7</span> 0.363   0.194 -<span style="color: #BB0000;">0.927</span> -<span style="color: #BB0000;">1.42</span>     5 Hba-a1            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MSigDB__Granuloc… 0.125  0.363   0.154 -<span style="color: #BB0000;">0.896</span> -<span style="color: #BB0000;">1.38</span>     5 S100a9,Mal…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MSigDB__Granuloc… 0.152  0.363   0.121 -<span style="color: #BB0000;">0.829</span> -<span style="color: #BB0000;">1.35</span>    13 S100a9,Mal…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MSigDB__Hematopo… 0.094<span style="text-decoration: underline;">0</span> 0.363   0.167 -<span style="color: #BB0000;">0.889</span> -<span style="color: #BB0000;">1.42</span>     8 S100a9,Mal…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MSigDB__Late_pro… 0.172  0.363   0.125 -<span style="color: #BB0000;">0.847</span> -<span style="color: #BB0000;">1.32</span>     6 S100a9,S10…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MSigDB__Macropha… 0.153  0.363   0.128 -<span style="color: #BB0000;">0.847</span> -<span style="color: #BB0000;">1.35</span>     8 S100a9,Ltf…       0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MSigDB__Mature_a… 0.175  0.363   0.124 -<span style="color: #BB0000;">0.843</span> -<span style="color: #BB0000;">1.31</span>     6 Malat1            0</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 23 more rows</span></span></span></code></pre>
<p>As shown, the output is a table with following variables-
<em>pathaway</em>: defines a set of markers used for GSEA analysis /
<em>pval</em>: p-value / <em>padj</em>: adjusted p-value /
<em>log2err</em>: the expected error for the standard deviation of the
p-value logarithm / <em>ES</em>: enrichment score / <em>NES</em>:
normalized enrichment score /<em>size</em>: size of the pathway after
removing genes not present in ‘names(stats)’ / <em>leadingEdge</em>:
vector with indexes of leading edge genes that drive the enrichment
/<em>cluster</em>: cluster number.</p>
<p><strong>Note:</strong> As you can see in the plot above, the results
in cluster 2 have very low significance in terms of adjusted p-value. In
this case it might be useful to plot <em>NES</em> values instead. Let’s
have a look how to do that.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the top 5 most significant results</span></span>
<span><span class="fu"><a href="../reference/plotMarkerGSEA.html">plotMarkerGSEA</a></span><span class="op">(</span><span class="va">GSEAres</span>, topN <span class="op">=</span> <span class="fl">5</span>, plotNES <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/plot-NES-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="custom-settings">Custom settings<a class="anchor" aria-label="anchor" href="#custom-settings"></a>
</h2>
<p>The functions presented here provide multiple options how to run
analyses and plot the results. Additionally, users can provide their own
sets of cell type / tissue markers (these do not have to be
bone-related). Furthermore, the outputs of the plotting function are
editable <em>ggplot</em> objects. In this section we will cover both of
these options to customize provided functions.</p>
<div class="section level3">
<h3 id="a--provide-custom-set-of-markers">a. Provide custom set of markers<a class="anchor" aria-label="anchor" href="#a--provide-custom-set-of-markers"></a>
</h3>
<p>As part pf the package we provide a set of bone-related markers. The
inputs to overrepresentation analysis vs. GSEA analysis have slightly
different format (<em>data.frame</em> vs. <em>list</em>, respectively).
So let’s have a look at what format are the inputs and how do we provide
those to both functions from this package.</p>
<div class="section level4">
<h4 id="a-1-markers-for-overrepresentation-analysis">a.1 Markers for overrepresentation analysis<a class="anchor" aria-label="anchor" href="#a-1-markers-for-overrepresentation-analysis"></a>
</h4>
<p>The marker database provided in this package for overrepresentation
analysis can be accessed in following way:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access set of markers</span></span>
<span><span class="va">markerDatabase</span> <span class="op">=</span> <span class="va">markerDatabase</span></span>
<span></span>
<span><span class="co"># view variables in the markerDatabase data.frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markerDatabase</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   tissue              cellType       gene          database</span></span>
<span><span class="co">#&gt; 1   Bone            Osteoclast       Acp5 CellMarker2_mouse</span></span>
<span><span class="co">#&gt; 2   Bone       Osteogenic cell Galectin-3 CellMarker2_mouse</span></span>
<span><span class="co">#&gt; 3   Bone            Osteoclast       Ctsk CellMarker2_mouse</span></span>
<span><span class="co">#&gt; 4   Bone            Osteoclast        Ctr CellMarker2_mouse</span></span>
<span><span class="co">#&gt; 5   Bone            Osteoclast       Mmp9 CellMarker2_mouse</span></span>
<span><span class="co">#&gt; 6   Bone Mesenchymal stem cell       Cd45 CellMarker2_mouse</span></span></code></pre>
<p>If you want to provide your own set of markers, the
<em>data.frame</em> must have all following columns: <em>tissue,
cellType, gene, database</em>. Each row must contain only one gene
(i.e. marker). Let’s now have a look how to use the markerDatabase
object in the overrepresentation analysis with
<code>markerEnrich</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run overrepresentation analysis uasing top 10 variable markers (features) per cluster</span></span>
<span><span class="va">ORres2</span> <span class="op">=</span> <span class="fu"><a href="../reference/markerEnrich.html">markerEnrich</a></span><span class="op">(</span>varMarkers <span class="op">=</span> <span class="va">varMarkers</span>, markerDB <span class="op">=</span> <span class="va">markerDatabase</span>, topN <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the top 5 most significant results</span></span>
<span><span class="fu"><a href="../reference/plotMarkerEnrich.html">plotMarkerEnrich</a></span><span class="op">(</span><span class="va">ORres2</span>, topN <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/overrepCutomMarkers-1.png" width="576"></p>
<p>Using the input in this way allows us to for example filter out
markers coming from sources we might not trust as much. Let’s say we
would like to omit markers coming from <em>CellMarker2_mouse_sc</em>
database (source). Let’s have a look how to do that.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter out markers originating from CellMarker2_mouse source</span></span>
<span><span class="va">filtDB</span> <span class="op">=</span> <span class="va">markerDatabase</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">database</span> <span class="op">!=</span> <span class="st">"CellMarker2_mouse_sc"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># pass the filtered set of markers to overrepresentation analysis</span></span>
<span><span class="va">ORresFilt</span> <span class="op">=</span> <span class="fu"><a href="../reference/markerEnrich.html">markerEnrich</a></span><span class="op">(</span>varMarkers <span class="op">=</span> <span class="va">varMarkers</span>, markerDB <span class="op">=</span> <span class="va">filtDB</span>, topN <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the top 5 most significant results</span></span>
<span><span class="fu"><a href="../reference/plotMarkerEnrich.html">plotMarkerEnrich</a></span><span class="op">(</span><span class="va">ORresFilt</span>, topN <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/overrepFilter-1.png" width="576"></p>
<p>As you can see, cell types from <em>CellMarker2_mouse_sc</em> are not
reported in the results, since the markers were filtered out.</p>
</div>
<div class="section level4">
<h4 id="a-2-markers-for-gsea-analysis">a.2 Markers for GSEA analysis<a class="anchor" aria-label="anchor" href="#a-2-markers-for-gsea-analysis"></a>
</h4>
<p>Marker input to the GSEA analysis is in form of a named
<em>list</em>. The names correspond to a given set and the content of
the given set is a string vector containing given markers. There are 4
marker sets included in this package, let’s have a look at them.</p>
<ol style="list-style-type: decimal">
<li><strong>List with cell type specific markers merged from multiple
source databases</strong></li>
</ol>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access set of cell type markers (merged from multiple source databases)</span></span>
<span><span class="va">cellType_list</span> <span class="op">=</span> <span class="va">cellType_list</span></span>
<span></span>
<span><span class="co"># view contents (maximum 6 markers per sublist)</span></span>
<span><span class="fu"><a href="../reference/headList.html">headList</a></span><span class="op">(</span><span class="va">cellType_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $Adipocyte</span></span>
<span><span class="co">#&gt; [1] "Apoe"   "Lpl"    "Adipoq" "Cebpb"  "Pparg"  "Cebpa" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Adult stem cell`</span></span>
<span><span class="co">#&gt; [1] "Cd34" "Ly6a" "Thy1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Alternatively activated macrophage`</span></span>
<span><span class="co">#&gt; [1] "Arg1"   "Retnla"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Anti-inflammatory macrophage`</span></span>
<span><span class="co">#&gt; [1] "Cx3cr1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Antibody-secreting cell`</span></span>
<span><span class="co">#&gt; [1] "Cd81"  "Il6st" "Cd39"  "Epcam"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Antigen-presenting cell`</span></span>
<span><span class="co">#&gt; [1] "Cd83"   "Ccr7"   "H2-Ab1" "Ptprc"  "H2-Eb1" "Mhc Ii"</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li><strong>List with tissue specific markers merged from multiple
source databases</strong></li>
</ol>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access set of tissue markers (merged from multiple source databases)</span></span>
<span><span class="va">tissue_list</span> <span class="op">=</span> <span class="va">tissue_list</span></span>
<span></span>
<span><span class="co"># view contents (maximum 6 markers per sublist)</span></span>
<span><span class="fu"><a href="../reference/headList.html">headList</a></span><span class="op">(</span><span class="va">tissue_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $Bone</span></span>
<span><span class="co">#&gt; [1] "Acp5"       "Galectin-3" "Ctsk"       "Ctr"        "Mmp9"      </span></span>
<span><span class="co">#&gt; [6] "Cd45"      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Bone marrow`</span></span>
<span><span class="co">#&gt; [1] "Flt3"   "Cr2"    "Csf1r"  "Adgre1" "Fcer1g" "Fcgr1" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Bone or Chondrocyte`</span></span>
<span><span class="co">#&gt; [1] "Twist2"        "Tnfaip6"       "Gm13652"       "Alx4"         </span></span>
<span><span class="co">#&gt; [5] "a"             "2700069I18Rik"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Chondrocyte</span></span>
<span><span class="co">#&gt; [1] "Pi15"   "Kcnt2"  "Fmo1"   "Rgs5"   "Atp1a2" "Lamc3"</span></span></code></pre>
<ol start="3" style="list-style-type: decimal">
<li><strong>List with cell type specific markers separated by source
database</strong></li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access set of cell type markers separated by source</span></span>
<span><span class="va">cellTypeSource_list</span> <span class="op">=</span> <span class="va">cellTypeSource_list</span></span>
<span></span>
<span><span class="co"># view contents (maximum 6 markers per sublist)</span></span>
<span><span class="fu"><a href="../reference/headList.html">headList</a></span><span class="op">(</span><span class="va">cellTypeSource_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $CellMarker2_mouse__Adipocyte</span></span>
<span><span class="co">#&gt; [1] "Apoe"   "Lpl"    "Adipoq" "Cebpb"  "Pparg"  "Cebpa" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse__Adult stem cell`</span></span>
<span><span class="co">#&gt; [1] "Cd34" "Ly6a" "Thy1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse__Alternatively activated macrophage`</span></span>
<span><span class="co">#&gt; [1] "Arg1"   "Retnla"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse__Anti-inflammatory macrophage`</span></span>
<span><span class="co">#&gt; [1] "Cx3cr1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse__Antibody-secreting cell`</span></span>
<span><span class="co">#&gt; [1] "Cd81"  "Il6st" "Cd39"  "Epcam"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse__Antigen-presenting cell`</span></span>
<span><span class="co">#&gt; [1] "Cd83"   "Ccr7"   "H2-Ab1" "Ptprc"  "H2-Eb1" "Mhc Ii"</span></span></code></pre>
<ol start="4" style="list-style-type: decimal">
<li><strong>List with cell type specific markers merged from multiple
source databases</strong></li>
</ol>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access set of tissue markers separated by source</span></span>
<span><span class="va">tissueSource_list</span> <span class="op">=</span> <span class="va">tissueSource_list</span></span>
<span></span>
<span><span class="co"># view contents (maximum 6 markers per sublist)</span></span>
<span><span class="fu"><a href="../reference/headList.html">headList</a></span><span class="op">(</span><span class="va">tissueSource_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $CellMarker2_mouse__Bone</span></span>
<span><span class="co">#&gt; [1] "Acp5"       "Galectin-3" "Ctsk"       "Ctr"        "Mmp9"      </span></span>
<span><span class="co">#&gt; [6] "Cd45"      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse__Bone marrow`</span></span>
<span><span class="co">#&gt; [1] "Flt3"   "Cr2"    "Csf1r"  "Adgre1" "Fcer1g" "Fcgr1" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $CellMarker2_mouse_sc__Bone</span></span>
<span><span class="co">#&gt; [1] "Thy1"    "Fap"     "Pdpn"    "Col22a1" "Lgr5"    "Ngp"    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`CellMarker2_mouse_sc__Bone marrow`</span></span>
<span><span class="co">#&gt; [1] "Pdcd1"  "Sell"   "Ly6a"   "Ifng"   "Arg1"   "Cxcl10"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $MSigDB__Bone</span></span>
<span><span class="co">#&gt; [1] "Slc9a2"   "Camk1d"   "Tnks1bp1" "S100a11"  "Tmub1"    "Spon2"   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`MSigDB__Bone marrow`</span></span>
<span><span class="co">#&gt; [1] "Ptma"   "Oaz1"   "Xpot"   "Itk"    "Ubb"    "Ubb-ps"</span></span></code></pre>
<p>Now let’s pass a selected set of markers to the markerGSEA function.
<strong>Note:</strong> once you provide your list of markers to the
<em>markers</em> variable, <em>systemLevel</em> and <em>byDatabase</em>
options will not do anything anymore. You have to create your list
according to your needs.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run GSEA with your set of markers</span></span>
<span><span class="va">GSEAcellTypeSource</span> <span class="op">=</span> <span class="fu"><a href="../reference/markerGSEA.html">markerGSEA</a></span><span class="op">(</span>varMarkers <span class="op">=</span> <span class="va">varMarkers</span>, markers <span class="op">=</span> <span class="va">cellTypeSource_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot NES of the top 5 most significant results</span></span>
<span><span class="fu"><a href="../reference/plotMarkerGSEA.html">plotMarkerGSEA</a></span><span class="op">(</span><span class="va">GSEAcellTypeSource</span>, topN <span class="op">=</span> <span class="fl">5</span>, plotNES <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/marker-GSEA-custom-1.png" width="576"></p>
</div>
</div>
<div class="section level3">
<h3 id="b--edit-generated-plots">b. Edit generated plots<a class="anchor" aria-label="anchor" href="#b--edit-generated-plots"></a>
</h3>
<p>The outputs of the plotting functions are editable <em>ggplot</em>
objects that you can adjust to your liking. Let’s have a look how to do
that.</p>
<p>First let’s generate the default plot.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run overrepresentation analysis uasing top 10 variable markers (features) per cluster</span></span>
<span><span class="va">ORres</span> <span class="op">=</span> <span class="fu"><a href="../reference/markerEnrich.html">markerEnrich</a></span><span class="op">(</span><span class="va">varMarkers</span>, topN <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the top 5 most significant results</span></span>
<span><span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotMarkerEnrich.html">plotMarkerEnrich</a></span><span class="op">(</span><span class="va">ORres</span>, topN <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/marker-overrep-1.png" width="576"></p>
<p>Let’s now change the color scheme and add a plot title.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/" class="external-link">viridisLite</a></span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">=</span> <span class="va">p</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"D"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Bone cell type overrepresentation analysis"</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="intro_to_BoneCellType_files/figure-html/edit%20plot-1.png" width="576"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Kristyna Kupkova.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
